src/quest.cpp

```diff
@@ -686,31 +711,49 @@ void new_quest(struct char_data *mob, bool force_assignation=FALSE)
     snprintf(buf, sizeof(buf), "Stripping Johnson status from %s (%ld) due to mob not having any quests to assign.",
             GET_NAME(mob), GET_MOB_VNUM(mob));
     mudlog(buf, NULL, LOG_SYSLOG, true);
-    return;
+    return -1;
   }
-
-  for (i = 0;;)
-  {
-    bool allow_disconnected = vnum_from_non_connected_zone(quest_table[i].johnson);
-    if (quest_table[i].johnson == GET_MOB_VNUM(mob) &&
-        (force_assignation || (!number(0, num - 1) 
-                               && !(num > 1 
-                                    && GET_SPARE2(mob) == i 
-                                    && (allow_disconnected 
-                                        || !vnum_from_non_connected_zone(quest_table[i].vnum)))))) {
-      GET_SPARE2(mob) = i;
-      return;
+  
+  // Build array of quests for this johnson, excluding quests that are already
+  // done or max_rep is below character rep. We include those with min_rep
+  // higher than character rep because we want johnsons to hint to available
+  // runs at higher character rep.
+  for (i = 0;i < top_of_questt;i++) {
+          bool allow_disconnected = vnum_from_non_connected_zone(quest_table[i].johnson);
+          if (quest_table[i].johnson == GET_MOB_VNUM(mob)
+          && (allow_disconnected || !vnum_from_non_connected_zone(quest_table[i].vnum))) {
+            if (GET_REP(ch) > quest_table[i].max_rep) {
+                continue;
+            }
+            
+            bool found = FALSE;
+            for (int q = QUEST_TIMER - 1; q >= 0; q--) {
+              if (GET_LQUEST(ch, q) == quest_table[i].vnum) {
+                  found = TRUE;
+                  break;
+              }
+            }
+            if (found) {
+                continue;
+            } else {
+                temp_entry.index = i;
+                temp_entry.rep = quest_table[i].min_rep;
+                qlist.push_back(temp_entry);
+            }
     }
-    if ((i + 1) <= top_of_questt)
-      i++;
-    else
-      i = 0;
+}
+  // Sort vector by reputation and return a quest if vector is not empty.
+  if (!qlist.empty()) {
+    sort(qlist.begin(), qlist.end(), [] (const quest_entry &a, const quest_entry &b) { return a.rep < b.rep; });
```

Comment:
Both the warning and the fact that lambda function on `sort()` won't compile, means you're compiling with C++03 instead of C++11. Both are legal in C++11 (the local type restriction has been lifted). While testing G++ days ago, compiler claimed there was C++11 code in the project, so I went ahead and assumed using C++11 is ok. Do you want me to rewrite it to adhere to C++03?

Reply:
Nah we can use C++11 as long as it works properly when compiled that way. Seems like it does, if you were able to test it with the others.

Reply:
Can you share the changes you made to the makefile for C++11? My version's fighting me, the clang++ I get through Homebrew is resistant to change when coupled with the existing makefile.

Reply:
I didn't actually change anything on that regard as clang++12 on Fedora works like that as default. Try using the `-std=c++11` switch or even better `-std=c++1y` if it will accept it. Which version of clang++ is it by the way? I don't have a mac available.

Reply:
```
$ clang++ --version                                                                                                                                                                    (master)
Apple clang version 13.0.0 (clang-1300.0.29.3)
Target: arm64-apple-darwin21.1.0
```

Doesn't work with that flag-- the c++11 code compiles, but then the sockets library appears to have changed enough to break the compile:

```
clang++ -std=c++11  -DUSE_MEMORY_CANARIES -DDEBUG -DSELFADVANCE -Dosx -I /usr/local/include/ -Wall -Wextra -Wno-unused-parameter -ggdb  -c -o comm.o comm.cpp
comm.cpp:498:53: error: invalid operands to binary expression ('__bind<int &, sockaddr *, unsigned long>' and 'int')
  if (bind(s, (struct sockaddr *) & sa, sizeof(sa)) < 0) {
      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/system_error:305:1: note: candidate function not viable: no known conversion from '__bind<int &, sockaddr *, unsigned long>' to 'const std::error_condition' for 1st argument
operator<(const error_condition& __x, const error_condition& __y) _NOEXCEPT

<about 10 other potential operators snipped for brevity>
```

For now, let's revert to C++03 so I can get this compiled and tested on my machine before the vacation.

Reply:
https://github.com/luciensadi/AwakeMUD/pull/582/commits/f60ed75a9d3f7ae6c282e3974a415e40f3f0bb90 changes the lambda function to a normal comparator function and moved the template struct for the vector to quest.h so it's not local anymore. This should compile properly now BUT when I switched `-std=c++03` on in order to test, compilation failed on my end due to other C++11 requiring code. This leads me to believe that something fishy it's going on with that Clang version of yours or maybe it's something in Clang13. The changes should allow you to compile though since you have no other issue with apart from this one. I will update my server to Fedora 35 today (I needed to do this anyway) which comes with Clang13 and see if this is reproducible on my end and what's actually wrong with code that should be compliant but isn't. It's better if we use the same version going forward, anyhow.